### PR TITLE
[CARBONDATA-2625] Optimize the performance of CarbonReader read many files

### PR DESCRIPTION
**Pull request by rahulforallp**
REf : https://github.com/apache/carbondata/pull/2391

About the issue: it's timeout and no result in 8 minutes when read more than 10 million data with 140 files, Even though increase 200000 rows for each carbon Writer and it can reduce the index files and data files when the number of rows is 13000000, but when there are more than 1 billion or more, the number of files still still many. I check the code and find read more 140 files can be optimize:

In the cache.getAll, the IO is more than 140 if there are 140 carbon files, in fact, the IO are more than 70 * 140 times, it's slow and can be optimized

Secondly, there are some duplicate operate in getDataMaps and can be optimized

Thirdly, SDK need much time to create multiple carbonRecorderReader, it need more than 8 minutes by testing 150 files and 15million rows data when create more than 16 carbonReorederReader if the machine has 8 cores . It can be optimized

By optimizing the three points,including cache.getAll, getDatamaps and create carbonRecordReader, now SDK can work for reading 150 files and 15million rows data in 8 minutes, it need about 340 seconds by testing.

One case: 150 files , each file has 200000 rows, total rows is 15000000
Finished write data time: 449.102 s
Finished build reader time:192.596 s
Read first row time: 192.597 s, including build reader
Read time:341.556 s, including build reader

Another case: 15 files , each file has 2000000 rows, total rows is 15000000
Finished write data time: 286.907 s
Finished build reader time: 134.665 s
Read first row time: 134.666 s, including build reader
Finished read, the count of rows is:15000000
Read time:156.427 s, including build reader

Be sure to do all of the following checklist to help us incorporate
your contribution quickly and easily:

    Any interfaces changed?
    Yes, add new one for optimizing performance

    Any backward compatibility impacted?
    NA

    Document update required?
    NO

    Testing done
    add example for it

    For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.
    NO

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6702/

**Comment by rahulforallp** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5570/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5555/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5573/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5556/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5574/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5559/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5575/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6729/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6731/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5563/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5577/

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 1]
```diff
@@ -121,6 +122,7 @@ public static CarbonReaderBuilder builder(String tablePath) {
    */
   public void close() throws IOException {
     validateReader();
+    SDKDetailQueryExecutor.shutdownThreadPool();
```
Comment (gvramana) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Static one we cannot shutdown, as another CarbonReader might be reading same or different table in same process

**Pull request closed without merging**